src/main/java/com/ibm/watson/modelmesh/payload/AsyncPayloadProcessor.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright 2023 IBM Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy
+ * of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.ibm.watson.modelmesh.payload;
+
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.slf4j.LoggerFactory;
+import org.slf4j.Logger;
+
+public class AsyncPayloadProcessor implements PayloadProcessor {
+
+    private static final Logger logger = LoggerFactory.getLogger(AsyncPayloadProcessor.class);
+
+    private final PayloadProcessor delegate;
+
+    private final FixedSizeConcurrentLinkedDeque<Payload> requestPayloads;
+    private final FixedSizeConcurrentLinkedDeque<Payload> responsePayloads;
+
+    public AsyncPayloadProcessor(PayloadProcessor delegate) {
+        this(delegate, 1, TimeUnit.MINUTES, Executors.newScheduledThreadPool(1), 1000);
+    }
+
+    public AsyncPayloadProcessor(PayloadProcessor delegate, int delay, TimeUnit timeUnit,
+                                 ScheduledExecutorService executorService, int capacity) {
+        this.delegate = delegate;
+        this.requestPayloads = new FixedSizeConcurrentLinkedDeque<>(capacity);
+        this.responsePayloads = new FixedSizeConcurrentLinkedDeque<>(capacity);
+
+        executorService.scheduleWithFixedDelay(() -> {
+            Payload p;
+            while ((p = requestPayloads.poll()) != null) {
+                delegate.processRequest(p);
+                maybeRelease(p);
+            }
+            int droppedRequest = requestPayloads.dropped.getAndSet(0);
+            if (droppedRequest > 0) {
+                logger.warn("{} request payloads were skipped because of {} capacity limit", droppedRequest, capacity);
+            }
+            while ((p = responsePayloads.poll()) != null) {
+                delegate.processResponse(p);
+                maybeRelease(p);
+            }
+            int droppedResponse = responsePayloads.dropped.getAndSet(0);
+            if (droppedResponse > 0) {
+                logger.warn("{} response payloads were skipped because of {} capacity limit", droppedResponse, capacity);
+            }
+        }, 0, delay, timeUnit);
+    }
+
+    private static void maybeRelease(Payload p) {
+        if (p.getData() != null) {
+            p.getData().release();
+        }
+    }
```

Comment:
while poking around with MM (in tests mostly) I did see some `null` `ByteBufs` in requests/response, that's the reason for the null check.

I'll see if I can refactor the whole PR moving retain/release within `Payload`.

Reply:
I think the only time a ByteBuf would be null is in an exception case. Which makes me think of another point - what should be logged if anything in the case of error responses? There's no actual payload but I guess the error code etc. could be recorded so that it can be correlated with the request payload.

Reply:
the failing status code is logged by `RemotePayloadProcessor`, e.g.:
```
11:17:05.420 [pool-7-thread-1](69) WARN  c.i.w.m.p.RemotePayloadProcessor - Processing Payload{id='84-1', modelId='myModel', method='mmesh.ExamplePredictor/predict', kind=request, metadata=Metadata(content-type=application/grpc,user-agent=grpc-java-netty/1.53.0,mm-model-id=myModel,another-custom-header=custom-value,grpc-accept-encoding=gzip,grpc-timeout=1999479u), data=4B} with request {data=EgJhYQ==, modelid=myModel, kind=request, id=84-1} didn't succeed: (POST http://localhost:8080/consumer/kserve/v2) 500 

```

Reply:
so I'd keep the possibility of having a 'null` `ByteBuf` and make sure that is handled correctly within processors.